Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java

```diff
@@ -368,8 +371,154 @@ private void updateDriverStatus(Boolean status) {
+            waypoints.put(startLatLng);
+            waypoints.put(endLatLng);
+
+            payload.put("waypoints", waypoints);
+            payload.put("mode", "CAR");
+            payload.put("calcPoints", true);
+
+            OutputStream stream = connection.getOutputStream();
+            stream.write(payload.toString().getBytes());
+            connection.connect();
+            int respCode = connection.getResponseCode();
+            InputStreamReader respReader;
+            if ((respCode < 200 || respCode >= 300) && respCode != 302) {
+                respReader = new InputStreamReader(connection.getErrorStream());
+                BufferedReader in = new BufferedReader(respReader);
+                String inputLine;
+                while ((inputLine = in.readLine()) != null) {
+                    result.append(inputLine);
+                }
+                JSONObject errorPayload = new JSONObject(result.toString());
+            } else {
+                respReader = new InputStreamReader(connection.getInputStream());
+                BufferedReader in = new BufferedReader(respReader);
+                String inputLine;
+                while ((inputLine = in.readLine()) != null) {
+                    result.append(inputLine);
+                }
+                JSONObject res = (JSONObject) new JSONArray(String.valueOf(result)).get(0);
+                JSONArray pathLatLng = (JSONArray) res.get("points");
+
+                SharedPreferences sharedPreff = this.getSharedPreferences(this.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+                SharedPreferences.Editor editor = sharedPreff.edit();
+                editor.putString("RIDE_WAYPOINT", String.valueOf(pathLatLng));
+                editor.apply();
+            }
+
+        } catch (Exception e) {
+            Log.e("get route api exception", e.toString());
+        }
+    }
+
+    private void updateDeviation(double latitude, double longitude, String waypoints) throws JSONException {
+        try {
+
+            SharedPreferences sharedPreff = this.getSharedPreferences(this.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+            List<LatLng> LatLngPoints = new ArrayList<LatLng>();
+            JSONArray waypointArray = new JSONArray(waypoints);
+
+            for (int i = 0; i < waypointArray.length(); i++) {
+                JSONObject pt = waypointArray.getJSONObject(i);
+                LatLngPoints.add(new LatLng((Double) pt.get("lat"), (Double) pt.get("lon")));
+            }
+
+
+            int resultIndex = PolyUtil.locationIndexOnEdgeOrPath(new LatLng(latitude, longitude), LatLngPoints, PolyUtil.isClosedPolygon(LatLngPoints), true, 30.0);
+
+            if (resultIndex == -1) {
+                int deviation = Integer.parseInt(sharedPreff.getString("RIDE_WAYPOINT_DEVIATION_COUNT", null));
+                deviation = deviation + 1;
+                SharedPreferences.Editor editor = sharedPreff.edit();
```

Comment:
We have a function to update SharedPref variables in this file :- `updateStorage(k, v)` use that function here